xopt/generators/bayesian/bax/acquisition.py

```diff
@@ -0,0 +1,135 @@
+            self.xs_exe
+        )  # call model on some data to avoid errors. we don't need this result.
+
+        """
+        construct a batch of size n_samples fantasy models,
+        where each fantasy model is produced by taking the model
+        at the current iteration and conditioning it
+        on one of the sampled execution path subsequences:
+        """
+        xs_exe_transformed = self.model.input_transform(self.xs_exe)
+        ys_exe_transformed = self.model.outcome_transform(self.ys_exe)[0]
+        self.fmodels = self.model.condition_on_observations(
+            xs_exe_transformed, ys_exe_transformed
+        )
+
+    @t_batch_mode_transform(expected_q=1, assert_output_shape=False)
+    def forward(self, X: Tensor) -> Tensor:
+        r"""Evaluate Expected Information Gain on the candidate set X.
+
+        Args:
+            X: A `(b1 x ... bk) x 1 x d`-dim batched tensor of `d`-dim design points.
+                Expected Information Gain is computed for each point individually,
+                i.e., what is considered are the marginal posteriors, not the
+                joint.
+
+        Returns:
+            A `(b1 x ... bk)`-dim tensor of Expected Information Gain values at the
+            given design points `X`.
+        """
+
+        """
+        Use the current & fantasy models to compute a
+        Monte-Carlo estimate of the Expected Information Gain:
+        see https://arxiv.org/pdf/2104.09460.pdf:
+        eq (4) and the last sentence of page 7)
+        """
+        # calculcate the variance of the posterior for each input x
+        p = self.model.posterior(X)
+        var_p = p.variance
+        var_p = var_p.reshape(var_p.shape[:-1])
+
+        # calculcate the variance of the fantasy posteriors
+        pfs = self.fmodels.posterior(
+            (
+                X.reshape(*X.shape[:-2], 1, *X.shape[-2:]).expand(
+                    *X.shape[:-2], self.xs_exe.shape[0], *X.shape[-2:]
+                )
+            )
+        )
+        var_pfs = pfs.variance
+        var_pfs = var_pfs.reshape(var_pfs.shape[:-1])
+
+        # calculate Shannon entropy for posterior given the current data
+        h_current = 0.5 * torch.log(2 * torch.pi * var_p) + 0.5
+
+        # calculate the Shannon entropy for the fantasy posteriors
+        h_fantasies = 0.5 * torch.log(2 * torch.pi * var_pfs) + 0.5
+
+        # compute the Monte-Carlo estimate of the Expected value of the entropy
+        avg_h_fantasy = torch.mean(h_fantasies, dim=-2)
+
+        """
+        use the above entropies to compute the Expected Information Gain,
+        where the terms in the equation below correspond to the terms in
+        eq (4) of https://arxiv.org/pdf/2104.09460.pdf
+        (avg_h_fantasy is a Monte-Carlo estimate of the second term on the right)
+        """
+        eig = h_current - avg_h_fantasy
+
+        return eig.reshape(X.shape[:-2])
+
+
+class PosteriorUncertainty(AnalyticAcquisitionFunction):
```

Comment:
this should be defined in another file, please grab the qPosteriorVariance class definition from bayesian_exploration.py and put it in a separate file in custom_botorch/